### ISSUE TITLE
Diff via libgit2 on bare repo almost 10 times slower than git

### ISSUE THREAD

**Issue by arthurschreiber** [libgit2/libgit2#3920]
I believe that generating a diff via libgit2 is considerably slower than via git.

I've encountered this while using rugged, but I'm able to reproduce the "bad performance" I've seen in rugged using the libgit2 `diff` example.

Here's a "simple way" to reproduce this:

```
arthurschreiber$ git clone https://github.com/Seeed-Studio/Grove_LED_Matrix_Driver --bare

Cloning into bare repository 'Grove_LED_Matrix_Driver.git'...
remote: Counting objects: 540, done.
remote: Total 540 (delta 0), reused 0 (delta 0), pack-reused 540
Receiving objects: 100% (540/540), 5.37 MiB | 910.00 KiB/s, done.
Resolving deltas: 100% (264/264), done.
Checking connectivity... done.

arthurschreiber$ time libgit2/build/examples/diff --git-dir Grove_LED_Matrix_Driver.git 9bdc746c1d1999c9a073cd2839899b84af2ff0cd fdb7257222e63a9b241b3899ff426bcd387b3821 >/dev/null

real	0m1.551s
user	0m1.520s
sys	0m0.023s

arthurschreiber$ time git --git-dir Grove_LED_Matrix_Driver.git diff 9bdc746c1d1999c9a073cd2839899b84af2ff0cd fdb7257222e63a9b241b3899ff426bcd387b3821 >/dev/null

real	0m0.153s
user	0m0.125s
sys	0m0.018s

```

---

/cc @ethomson @brianmario

**Comment by arthurschreiber** [comment]
One more thing I missed:

```
arthurschreiber$ git --version
git version 2.9.0
```

**Comment by pks-t** [comment]
This is expected in this case, as in contrast to git.git we generate a diff for the binary file `Firmware/grove_led_matrix_mbed/build/MBED_A1.axf` and other binary files, which requires roughly 95% of the time generating the diff. When you enable binary diffs for git.git with `git diff --binary`, you get roughly the same timings.

**Comment by arthurschreiber** [comment]
> This is expected in this case, as in contrast to git.git we generate a diff for the binary file Firmware/grove_led_matrix_mbed/build/MBED_A1.axf and other binary files, which requires roughly 95% of the time generating the diff

Why are binary diffs generated in libgit2 by default? I see we have a `GIT_DIFF_SHOW_BINARY` flag, which I understand would trigger the generation of the binary diff, but it's never set in the example code, and nowhere else in libgit2.

Am I missing something?

**Comment by pks-t** [comment]
No, you're correct. It is certainly our shortcoming if we calculate binary diffs even though `GIT_DIFF_SHOW_BINARY` is not set.

**Comment by pks-t** [comment]
Hum. I think the actual problem is that we deflate complete binary blobs only to show `Binary files a/foo/bar and b/foo/bar differ`. So when we only want to output that files _do_ in fact differ, without printing their diffs, it will usually be enough to short-circuit this by e.g. looking at the blob's sizes without decompressing them or somesuch heuristics.

**Comment by arthurschreiber** [comment]
We already know that blobs differ if they have different blob oids.

**Comment by arthurschreiber** [comment]
> Hum. I think the actual problem is that we deflate complete binary blobs only to show Binary files a/foo/bar and b/foo/bar differ.

Yeah, but don't we need to load the blob before we can check whether it's binary or not?

**Comment by pks-t** [comment]
Something as simple as the following fixes this use case without breaking any tests.

```
diff --git a/src/patch_generate.c b/src/patch_generate.c
index 927786a..67a13b7 100644
--- a/src/patch_generate.c
+++ b/src/patch_generate.c
@@ -349,20 +349,24 @@ static int diff_binary(git_patch_generated_output *output, git_patch_generated *
 		new_len = patch->nfile.map.len;
 	int error;
 
-	/* Create the old->new delta (as the "new" side of the patch),
-	 * and the new->old delta (as the "old" side)
-	 */
-	if ((error = create_binary(&binary.old_file.type,
-			(char **)&binary.old_file.data,
-			&binary.old_file.datalen,
-			&binary.old_file.inflatedlen,
-			new_data, new_len, old_data, old_len)) < 0 ||
-		(error = create_binary(&binary.new_file.type,
-			(char **)&binary.new_file.data,
-			&binary.new_file.datalen,
-			&binary.new_file.inflatedlen,
-			old_data, old_len, new_data, new_len)) < 0)
-		return error;
+	/* Only load contents if the user actually wants to diff
+	 * binary files. */
+	if (patch->base.diff_opts.flags & GIT_DIFF_SHOW_BINARY) {
+		/* Create the old->new delta (as the "new" side of the patch),
+		 * and the new->old delta (as the "old" side)
+		 */
+		if ((error = create_binary(&binary.old_file.type,
+				(char **)&binary.old_file.data,
+				&binary.old_file.datalen,
+				&binary.old_file.inflatedlen,
+				new_data, new_len, old_data, old_len)) < 0 ||
+			(error = create_binary(&binary.new_file.type,
+				(char **)&binary.new_file.data,
+				&binary.new_file.datalen,
+				&binary.new_file.inflatedlen,
+				old_data, old_len, new_data, new_len)) < 0)
+			return error;
+	}
 
 	error = giterr_set_after_callback_function(
 		output->binary_cb(patch->base.delta, &binary, output->payload),
```

**Comment by pks-t** [comment]
No, we do not actually have to load the blob. We have a heuristic which automatically sets a blob as being binary as soon as it exceeds a certain size.

**Comment by arthurschreiber** [comment]
I just tried this:

```
arthurschreiber$ time libgit2/build/examples/diff --git-dir Grove_LED_Matrix_Driver.git 9bdc746c1d1999c9a073cd2839899b84af2ff0cd fdb7257222e63a9b241b3899ff426bcd387b3821 >/dev/null

real	0m0.181s
user	0m0.154s
sys	0m0.019s
```

Much better! 👍 Thanks for looking into this @pks-t ❤️

**Comment by arthurschreiber** [comment]
Fixed by #3922. 👍
